### PR TITLE
Added Unit testing for all vue-vocabulary package

### LINKED ISSUE #962: npm ERR! missing script: test:unit
Our CI pipeline is failing with the following error:

> Run npm run lint
> npm ERR! missing script: test:unit

## Task

- [ ] create or restore the `unit` command so unit tests will run during pull requests
- [ ] the `test:unit` command should run all unit tests across the `fonts`, `vocabulary`, and `vue-vocabulary` packages where relevant

### LINKED ISSUE #965: Restore project unit tests
Our project consists of three sub-modules:

- fonts
- vocabulary
- vue-vocabulary

Each of the submodules used to have corresponding tests:

- [fonts](https://github.com/creativecommons/vocabulary/tree/5e7f4331d62267a4332da2ca484d29b0575e6007/packages/fonts/tests)
- [vocabulary](https://github.com/creativecommons/vocabulary/tree/5e7f4331d62267a4332da2ca484d29b0575e6007/packages/vocabulary/tests)
- [vue-vocabulary](https://github.com/creativecommons/vocabulary/tree/5e7f4331d62267a4332da2ca484d29b0575e6007/packages/vue-vocabulary/tests)

The unit tests were invoked in relevant continuous integration scripts:

- [ci-fonts](https://github.com/creativecommons/vocabulary/blob/238a272adc1f942a43ca6d9d72f0ddabc08c9df4/.github/workflows/ci-fonts.yml#L74-L101)
- [ci-vocabulary](https://github.com/creativecommons/vocabulary/blob/238a272adc1f942a43ca6d9d72f0ddabc08c9df4/.github/workflows/ci-vocabulary.yml#L75-L102)
- [ci-vue-vocabulary](https://github.com/creativecommons/vocabulary/blob/238a272adc1f942a43ca6d9d72f0ddabc08c9df4/.github/workflows/ci-vue.yml#L74-L101)

## Task

- [ ] restore all testing code in a `tests/` directory in the root of this project
    - [ ] `tests/vue-vocabulary`
- [ ] ensure we have script(s) defined in `package.json` to invoke the test suites
    - [ ] `test:vue-vocabulary`
- [ ] update CI test runners to invoke restored tests
    - [ ] `ci-vue.yml` (rename to `ci-vue-vocabulary.yml`) 
- [ ] remove fonts tests, since `sass-jest` is no longer maintained 
    - [ ] delete `vocabulary/.github/workflows/ci-fonts.yml` 
    - [ ] delete `vocabulary/.github/workflows/ci-vocabulary.yml` 
- [ ] delete all unused scripts and code after refactoring tests

### PR DESCRIPTION
**Pull request by MuluhGodson**
## Fixes

Closes #965  by @brylie 
Closes #962  by @brylie 

## Description
Reinstated test for @vue-vocabulary package.

So, for contributing, each time a new core module is added, or an existing module is refactored, a test should be written for that module in the package and run using the unit's test.

To summarize, task completed include:
- [x]  added a test unit for Vue-vocabulary
- [x]  added test script for vue-vocabulary
- [x]  added a general test script
- [x]  updated README.md to reflect changes made 

## Technical details
Packages installed include:
```bash
    vue/cli-plugin-unit-jest": "~4.5.0
    jest": "^27.2.5
    vue/test-utils": "^1.0.3
```

## Tests
Run the following commands to test each package individually
```bash
npm run test:vue-vocabulary
```
To run a general test, run this command
```bash
npm run test:unit
```

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: packages/vue-vocabulary/tests/unit/Dummy.spec.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+describe('Dummy', () => {
+    it('matches string with string', () => {
+      expect('Hello World!').toContain('Hello')
+    })
+  })
```
Comment (brylie) [inline packages/vue-vocabulary/tests/unit/Dummy.spec.js]: Lint
```suggestion
  })

```

File: packages/vue-vocabulary/tests/unit/VCheckbox.spec.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    await wrapper.setProps({
+      value: true,
+      disabled: true
+    })
+    const checkbox = wrapper.find('input[type="checkbox"]')
+    expect(checkbox.element.checked).toBeTruthy()
+    await checkbox.element.click()
+    expect(wrapper.emitted().input).toBeFalsy()
+  })
+})
```
Comment (brylie) [inline packages/vue-vocabulary/tests/unit/VCheckbox.spec.js]: Lint
```suggestion
})

```

File: packages/vue-vocabulary/tests/unit/jest.config.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    'js',
+    'jsx',
+    'json',
+    'vue',
+  ],
+  testMatch: [
+    '**/tests/unit/**/*.spec.(js|jsx|ts|tsx)|**/__tests__/*.(js|jsx|ts|tsx)'
+  ],
+
+};
```
Comment (brylie) [inline packages/vue-vocabulary/tests/unit/jest.config.js]: Lint
```suggestion
};

```

File: scripts/test_unit.sh [hunk 4]
```diff
@@ -0,0 +1,4 @@
+echo "Testing all packages"
+npm run test:fonts
+npm run test:vocabulary
+npm run test:vue-vocabulary
```
Comment (brylie) [inline scripts/test_unit.sh]: lint
```suggestion
npm run test:vue-vocabulary

```

File: scripts/test_unit.sh [hunk 5]
```diff
@@ -0,0 +1,4 @@
+echo "Testing all packages"
+npm run test:fonts
+npm run test:vocabulary
```
Comment (brylie) [inline scripts/test_unit.sh]: Do we still need these tests?

File: scripts/test_unit.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+echo "Testing all packages"
```
Comment (brylie) [inline scripts/test_unit.sh]: We can probably delete this file, if there is only one test suite (`test:vocabulary`).

**Review by brylie — CHANGES_REQUESTED** [review]
Just a couple of minor lint issues.

**Comment by brylie** [comment]
Overall, the `*.scss` tests don't strike me as very meaningful. They are essentially diffing files, which is done as part of code review already.

**Review by brylie — CHANGES_REQUESTED** [review]
Let's move the project tests into a `/tests/` directory in the project root, rather than in `packages/<package-name>/`

File: scripts/test_unit.sh (hunk 5 above)
Comment (MuluhGodson) [reply scripts/test_unit.sh]: Okay I think I will remove them

File: scripts/test_unit.sh (hunk 6 above)
Comment (MuluhGodson) [reply scripts/test_unit.sh]: You mean `test:vue-vocabulary`
It's the test that doesn't check the `*.scss` files

File: packages/vue-vocabulary/tests/unit/Dummy.spec.js (hunk 1 above)
Comment (brylie) [inline packages/vue-vocabulary/tests/unit/Dummy.spec.js]: Check indentation.

**Review by brylie — CHANGES_REQUESTED** [review]

**Comment by MuluhGodson** [comment]
to clarify, moving project tests into a `/tests/` directory in the project root
will the structure now be `<root_folder>\tests\<package_name>\tests` ?

**Comment by MuluhGodson** [comment]
also, should I delete the tests folders for `fonts` and `vocabulary` since the commands will also be removed for now?

File: scripts/test_unit.sh (hunk 6 above)
Comment (brylie) [reply scripts/test_unit.sh]: I think we can delete the `scripts/test_unit.sh` since we only intend to keep `test:vue-vocabulary`, correct?

**Comment by brylie** [comment]
> also, should I delete the tests folders for fonts and vocabulary since the commands will also be removed for now?

Yes. Let's clean up any unused folders, code, etc. as we develop.

**Comment by brylie** [comment]
> will the structure now be `<root_folder>\tests\<package_name>\tests`?

I'm not sure we need to include the `<package-name>` in the path, since there is only one package (`vue-vocabulary`) that we will be testing. What do you suggest?

Also, the second `tests/` is redundant.

**Comment by MuluhGodson** [comment]
I think that is the best option. To avoid redundancy. If in future subsequent test are written for the other packages, then the structure can be updated then.

**Review by brylie — APPROVED** [review]

**Review by brylie — CHANGES_REQUESTED** [review]
Since this pull request intends to close #965, please ensure that all of the checklist items in #965 are complete here.

image

**Comment by MuluhGodson** [comment]
Hello @brylie 
A PR was merged yesterday that seperated the CI files. we no longer have a ci-fonts etc. we have just `ci-build.yml`, `ci-lint.yml` and `ci-test.yml`

**Comment by MuluhGodson** [comment]
also, you had recommended yesterday that since we have just a test for `vue-vocabulary` we should just place the test files in the root `test` folder, with no `<package_name>` sub-folders.

https://github.com/creativecommons/vocabulary/pull/978#issuecomment-942527511

**Comment by brylie** [comment]
@MuluhGodson, good points. The project structure has changed since the checklist was complete. I'm mainly concerned about deleting unused code and files. Are there any unused scripts or code remaining in the project related to this pull request?

File: tests/unit/Dummy.spec.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+describe('Dummy', () => {
```
Comment (brylie) [inline tests/unit/Dummy.spec.js]: We can probably delete this Dummy spec, unless you think it might be useful.

Comment (MuluhGodson) [reply tests/unit/Dummy.spec.js]: oh yes. I just used that for initial testing. it's deleted now.

**Comment by MuluhGodson** [comment]
- [x] restore testing code in a `tests/` directory in the root of this project
  - [x] tests/vue-vocabulary see https://github.com/creativecommons/vocabulary/pull/978#issuecomment-942527511
- [x] ensure we have script(s) defined in package.json to invoke test suites `npm run test:unit` and `npm run test:vue-vocabulary` while both test will do the same things now I made a decision to leave them because the `npm run test:unit` is used by the CI pipeline to run a general test and when future test are included, this command will be relevant and it is a good idea to have the `npm run test:vue-vocabulary` solely for running test for the `vue-vocabulary` if test for the others are subsequently added.
- [x] update CI test runners to invoke restored test
   - [x] `ci-vue.yml` A PR merge yesterday changed all test files to just `ci-test.yml`.
- [x] remove fonts test, since `sass-jest` is no longer maintained.
  - [x] delete `ci-fonts.yml` see https://github.com/creativecommons/vocabulary/pull/977
  - [x] delete `ci-vocabulary.yml` see https://github.com/creativecommons/vocabulary/pull/977
- [x] delete all unused scripts and code after refactoring

**Comment by brylie** [comment]
I just browsed the code on your branch and it seems fairly clean. So, I acknowledge that I was mistaken with the comment above. Thanks also for revising the checklist :smile: 

I did notice there are two `jest.config.js` files. Can we consolidate those files so there is only one Jest configuration?

**Review by brylie — APPROVED** [review]

**Comment by brylie** [comment]
In any case, I've approved this pull request. I'll just wait for your response about the double Jest configs.

- https://github.com/MuluhGodson/vocabulary/blob/unit_test/jest.config.js
- https://github.com/MuluhGodson/vocabulary/blob/unit_test/tests/unit/jest.config.js

**Comment by brylie** [comment]
It's going to be great to have our CI scripts up and running again :smiley: Thank you for your excellent work in this regard!

**Comment by MuluhGodson** [comment]
@brylie oh yes, removing the `jest.config.js` found in the `test` folder has no effect so i think it's a good idea to delete it. I think i used it when i still ran the test using `jest -c jest.config.js` but now the test are run using `vue-cli` so it's not needed any further.
but the `jest.config.js` in the <root> folder is necessary so that the `jest` test written for `vue-vocabulary` will be run with `vue-cli`

**Comment by MuluhGodson** [comment]
@brylie sorry I erroneously closed this with a comment. it's reopened now.

**Comment by brylie** [comment]
OK, I noticed you deleted the `jest.config.js`. Are there any config settings that should be merged into the remaining `jest.config.js`?

**Comment by brylie** [comment]
If not, I'll go ahead and merge this PR.

**Comment by MuluhGodson** [comment]
No. i think the configs that were in the deleted `jest.config.js` are not useful in this context. It was for a previous setup I experimented with but was not happy with the results.

**Pull request merged**